airbyte-integrations/connectors/destination-clickhouse-v2/src/main/kotlin/io/airbyte/integrations/destination/clickhouse_v2/client/ClickhouseAirbyteClient.kt

```diff
@@ -138,26 +133,66 @@ class ClickhouseAirbyteClient(
                 }
                 .toMap()
 
+        val clickhousePks: List<String> =
+            tableSchemaWithoutAirbyteColumns.filterNot { it.isNullable }.map { it.columnName }
+        val airbytePks: List<String> =
+            when (stream.importType) {
+                is Dedupe ->
+                    sqlGenerator.extractPks(
+                        (stream.importType as Dedupe).primaryKey,
+                        columnNameMapping
+                    )
+                else -> listOf()
+            }
+
         val columnChanges: AlterationSummary =
             getChangedColumns(
                 tableSchemaWithoutAirbyteColumns,
                 airbyteSchemaWithClickhouseType,
+                clickhousePks,
+                airbytePks,
+            )
+
+        if (columnChanges.hasApplicableAlterations()) {
+            execute(
+                sqlGenerator.alterTable(
+                    columnChanges,
+                    properTableName,
+                ),
             )
-        if (columnChanges.isEmpty()) {
-            log.info { "No changes detected for table $properTableName" }
-            return
         }
-        execute(
-            sqlGenerator.alterTable(
-                columnChanges,
-                properTableName,
-            ),
-        )
+
+        if (columnChanges.hasDedupChange) {
+            log.info {
+                "Detected deduplication change for table $properTableName, applying deduplication changes"
+            }
+            val tempTableName = tempTableNameGenerator.generate(properTableName)
+            execute(sqlGenerator.createNamespace(tempTableName.namespace))
+            execute(
+                sqlGenerator.createTable(
+                    stream,
+                    tempTableName,
+                    columnNameMapping,
+                    true,
+                ),
+            )
+            execute(
+                sqlGenerator.copyTable(
+                    columnNameMapping,
+                    properTableName,
+                    tempTableName,
+                ),
+            )
+            execute(sqlGenerator.exchangeTable(tempTableName, properTableName))
+            execute(sqlGenerator.dropTable(tempTableName))
+        }
```

Comment:
Because we don't have transactions, do we need to do anything on startup to resume where we left (or start over) off in case of a failure?